Bluetooth: gatt: Fix DB corruption when adding/removing service

arthur-proglove opened the pull request:
Fix an issue in the gatt_register routine where removing a service and
adding it again would cause the database to have non-ascending orderdered
handles numbers and lead to an incomplete service discovery.

Example with the following GATT DB:

0x0001 Service 1
0x0002 Char 1.1
0x0003 Service 2
0x0004 Char 2.1

If we remove Service 1 and add it again and since we always append a
new service at the end of the database we end up with the following
incorrect DB

0x0003 Service 2
0x0004 Char 2.1
0x0001 Service 1
0x0002 Char 1.1

Fixes: #17294

Fix: Go through the database and look for a place where to insert the new
service.
Signed-off-by: Arthur Crepin-Leblond <arthur.crepin@proglove.de>

Vudentz commented:
How about a little simpler solution such as:

https://gist.github.com/Vudentz/7f2e7e7091b515eade57a0820a1b6045

Btw, the original design never intended to insert at the middle we just start doing when re-registering but going thru the database to find out if a service fits a given position is a little overkill and we may endup with gaps anyway it usually much faster to just append to the end if the user has not requested a specific position.

arthur-proglove commented:
> How about a little simpler solution such as:
> 
> https://gist.github.com/Vudentz/7f2e7e7091b515eade57a0820a1b6045
> 
Looks good, to me. I will update my PR.

> Btw, the original design never intended to insert at the middle we just start doing when re-registering but going thru the database to find out if a service fits a given position is a little overkill and we may endup with gaps anyway it usually much faster to just append to the end if the user has not requested a specific position.

True, that's why my previous PR first idea was just to clear the handles of a service when removing it (or before inserting it) so that it would get new ones assigned (the last handle+1) and would added to the end of the DB.

arthur-proglove committed:
Bluetooth: gatt: Fix DB corruption when adding/removing service

Fix an issue in the gatt_register routine where removing a service and
adding it again would cause the database to have non-ascending
orderdered handles numbers and lead to an incomplete service discovery.

Fix: Go through the database and look for a place where to insert
the new service.

Signed-off-by: Arthur Crepin-Leblond <arthur.crepin@proglove.de>

Vudentz commented on subsys/bluetooth/host/gatt.c:
```diff
@@ -534,6 +534,32 @@ static const struct bt_gatt_attr *find_attr(uint16_t handle)
+
+	/* DB shall always have its service in ascending order */
+	SYS_SLIST_FOR_EACH_CONTAINER(&db, tmp, node) {
+		if (tmp->attrs[0]->handle > svc->attrs[0]->handle) {
```
You should probably build check these since I didn't and apparently it is not building correcly e.g. it should have been attrs[].handle, etc.

arthur-proglove replied:
oh thank you, my bad. I forgot to compile with CONFIG_BT_GATT_DYNAMIC_DB=y :-(.
It's fixed now!

arthur-proglove commented:
Fixed last handle old/new value.

joerchan reviewed (changes requested):
Looks good. Just a minor comment.

joerchan commented:
@jhedberg Can we merge this one?

jhedberg commented:
Is this something to be backported to 1.14, or only relevant to the master branch?

@Vudentz I suppose you're now ok with it?

joerchan commented:
@jhedberg The bt_gatt_service_unregister API is available in v1.14.0 so I think this is something that should be backported.

jhedberg commented:
> @jhedberg The bt_gatt_service_unregister API is available in v1.14.0 so I think this is something that should be backported.

Ok, I added the required label for that. @Vudentz please double-check that this is correct.

Vudentz commented:
> > @jhedberg The bt_gatt_service_unregister API is available in v1.14.0 so I think this is something that should be backported.
> 
> Ok, I added the required label for that. @Vudentz please double-check that this is correct.

Yep, that should be okay now.

Vudentz reviewed (approved):
LGTM

jhedberg commented:
@arthur-proglove @joerchan @Vudentz seems like the automated backport failed, i.e. someone will need to manually create the necessary PR. Also, it seems this is missing a matching bug report. That's a requirement for any PR that's submitted against the LTS branch.

jhedberg commented:
There's some GATT-handle related CI breakage happening currently in master:
```
***** Booting Zephyr OS build zephyr-v1.14.0-1989-gd21ee0c397ec *****
Running test suite test_gatt
===================================================================
starting test - test_gatt_register
[00:00:00.000,000] <err> bt_gatt: Unable to register handle 0x0001
[00:00:00.000,000] <err> bt_gatt: Unable to register handle 0x0004
PASS - test_gatt_register
===================================================================
starting test - test_gatt_unregister
[00:00:00.000,000] <err> bt_gatt: Unable to register handle 0x0001

    Assertion failed at /home/buildslave/src/github.com/zephyrproject-rtos/zephyr/tests/bluetooth/gatt/src/main.c:119: test_gatt_unregister: (bt_gatt_service_register(&test_svc) is true)
Test service re-registration failed
FAIL - test_gatt_unregister
===================================================================
starting test - test_gatt_foreach
[00:00:00.000,000] <err> bt_gatt: Unable to register handle 0x0001

    Assertion failed at /home/buildslave/src/github.com/zephyrproject-rtos/zephyr/tests/bluetooth/gatt/src/main.c:164: test_gatt_foreach: (bt_gatt_service_register(&test_svc) is true)
Test service registration failed
FAIL - test_gatt_foreach
```

The above was from CI of #17255 - I wonder if this PR is at fault? @Vudentz any thoughts?

Vudentz commented:
> There's some GATT-handle related CI breakage happening currently in master:
> 
> ```
> ***** Booting Zephyr OS build zephyr-v1.14.0-1989-gd21ee0c397ec *****
> Running test suite test_gatt
> ===================================================================
> starting test - test_gatt_register
> [00:00:00.000,000] <err> bt_gatt: Unable to register handle 0x0001
> [00:00:00.000,000] <err> bt_gatt: Unable to register handle 0x0004
> PASS - test_gatt_register
> ===================================================================
> starting test - test_gatt_unregister
> [00:00:00.000,000] <err> bt_gatt: Unable to register handle 0x0001
> 
>     Assertion failed at /home/buildslave/src/github.com/zephyrproject-rtos/zephyr/tests/bluetooth/gatt/src/main.c:119: test_gatt_unregister: (bt_gatt_service_register(&test_svc) is true)
> Test service re-registration failed
> FAIL - test_gatt_unregister
> ===================================================================
> starting test - test_gatt_foreach
> [00:00:00.000,000] <err> bt_gatt: Unable to register handle 0x0001
> 
>     Assertion failed at /home/buildslave/src/github.com/zephyrproject-rtos/zephyr/tests/bluetooth/gatt/src/main.c:164: test_gatt_foreach: (bt_gatt_service_register(&test_svc) is true)
> Test service registration failed
> FAIL - test_gatt_foreach
> ```
> 
> The above was from CI of #17255 - I wonder if this PR is at fault? @Vudentz any thoughts?

Yep, it seems to be failing at registration which is indeed, strange, will try to reproduce locally.

arthur-proglove commented:
Thanks @Vudentz for your hotfix.
What should be the next step now?

joerchan commented:
@jhedberg Ok, I'll create the necessary issue, and the backport PR.
